### PR TITLE
fix!: correct validation of VAPID keys

### PR DESCRIPTION
**Pull request by sw1tchdev**
Ref: #656
Hello, this PR fixed incorrect validation of VAPID keys, but it's a breaking change.

### CODE REVIEW

**Comment by sw1tchdev** [comment]
@marco-c Hello, please check this PR and ref issue, thanks 🙂

File: src/vapid-helper.js [hunk 1]
```diff
@@ -110,11 +116,17 @@ function validatePrivateKey(privateKey) {
...
-  if (privateKey.length !== 32) {
+  if (decodedPrivateKey.length !== 32) {
     throw new Error('Vapid private key should be 32 bytes long when decoded.');
   }
+
+  const encodedPrivateKey = urlBase64.encode(decodedPrivateKey);
+
+  if (privateKey !== encodedPrivateKey) {
+    throw new Error('Vapid private key must be a URL safe Base 64 (without "=")');
+  }
```
Comment (marco-c) [inline src/vapid-helper.js:129]: Same here

**Comment by marco-c** [comment]
Sorry for the delay, the patch looks good overall, just a nit.

File: src/vapid-helper.js [hunk 2]
```diff
@@ -164,7 +172,7 @@ function validateExpiration(expiration) {
  * @param  {string} publicKey       The VAPID public key.
  * @param  {string} privateKey      The VAPID private key.
  * @param  {string} contentEncoding The contentEncoding type.
- * @param  {integer} [expiration]   The expiration of the VAPID JWT.
+ * @param  {number} [expiration]   The expiration of the VAPID JWT.
```
Comment (marco-c) [inline src/vapid-helper.js:175]: Please revert this change, as it is unrelated (you can submit it in a separate PR).

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
 coverage/
 node_modules/
 test/output/
+.idea
```
Comment (marco-c) [inline .gitignore:4]: Please revert this change, as it is unrelated (you can submit it in a separate PR).

File: src/vapid-helper.js [hunk 4]
```diff
@@ -93,9 +93,13 @@ function validatePublicKey(publicKey) {
     + 'encoded string.');
   }
 
-  publicKey = urlBase64.decode(publicKey);
+  if (!urlBase64.validate(publicKey)) {
+    throw new Error('Vapid public key must be a URL safe Base 64 (without "=")');
+  }
+
+  const decodedPublicKey = urlBase64.decode(publicKey);
```
Comment (marco-c) [inline src/vapid-helper.js:100]: Can you keep the old code for this? So the diff is smaller

File: src/vapid-helper.js [hunk 5]
```diff
@@ -110,9 +114,13 @@ function validatePrivateKey(privateKey) {
     + 'encoded string.');
   }
 
-  privateKey = urlBase64.decode(privateKey);
+  if (!urlBase64.validate(privateKey)) {
+    throw new Error('Vapid private key must be a URL safe Base 64 (without "=")');
+  }
+
+  const decodedPrivateKey = urlBase64.decode(privateKey);
```
Comment (marco-c) [inline src/vapid-helper.js:121]: Same here

**Review by marco-c — APPROVED** [review]
Thanks!
